server/ports.go

```diff
@@ -0,0 +1,81 @@
+package server
+
+import (
+	"fmt"
+	"sync"
+)
+
+// portManager is responsible for maintaining segregated access to the port field
+// via the use of rwLock sync.RWMutex and mustRead sync.Mutex
+// rwLock establishes a standard read write mutex that allows consecutive reads and exclusive writes
+// mustRead forces MustGetPort to wait until port has a none 0 value
+type portManger struct {
+	port             int
+	afterPortChanged func(port int)
+	rwLock           *sync.RWMutex
+	mustRead         *sync.Mutex
+}
+
+// newPortManager returns a newly initialised portManager with a pre-Locked portManger.mustRead sync.Mutex
+func newPortManager(afterPortChanged func(int)) portManger {
+	pm := portManger{
+		afterPortChanged: afterPortChanged,
+		rwLock:           new(sync.RWMutex),
+		mustRead:         new(sync.Mutex),
+	}
+	pm.mustRead.Lock()
```

Comment:
Locking it right away first looked off to me, but reading your comments below, it seems like a good trick to block execution until a port has been set

Reply:
So, this is a little hacky, mutex doesn't perfectly fit this context, another type of sync struct is a `waitgroup`, but a `waitgroup` applies a cumulative number of deltas to wait, and `ResetPort` forces an additional `Add()` call, but `ResetPort` can be called `n` times before `MustGetPort` is called. This would add `n` wait deltas to the group, which can't be removed with a single `Done()` call.

Another approach could be preventing `ResetPort` from executing if there is a lock, but I don't know how to do this without using another mutex to guard the function.

I also considered resetting the whole `portManager`, but I've realised that there is no way to reset a mutex already blocking without crashing any routine waiting to acquire the lock.